add quantized instancenorm operator

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37802 quant: clean up normalization channels_last handling
* #37781 quant: add test util for testing off by one quantized tensors
* #37352 move quantization normalization layers to aten/src/ATen/native/quantized/cpu/
* **#36847 add quantized instancenorm operator**
* #36835 add quantized groupnorm operator
* #37947 quant: remove hypothesis and int32 from layernorm test

Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

vkuzo committed:
add quantized instancenorm operator

Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -274,6 +310,15 @@ static auto registry = torch::RegisterOperators().op(
     "float output_scale, "
     "int output_zero_point) -> Tensor",
     torch::RegisterOperators::options().kernel<QGroupNorm>(
+        DispatchKey::QuantizedCPU))
```
Can we use the new op registration API that recently landed?

supriyar commented on test/quantization/test_quantized.py:
```diff
@@ -1650,6 +1650,83 @@ def test_group_norm(self, batches, num_groups, channels_per_group,
+            num_diff_off_by_one = torch.sum((diff > 0) * (diff <= Y_scale))
+            pct_diff_off_by_one = float(num_diff_off_by_one) / (diff.numel() + 1e-5)
+
+            note("LayerNorm failed:\n {} input vs\n {} actual vs \n{} expected"
```
nit: message

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -1657,6 +1657,83 @@ def test_group_norm(self, batches, num_groups, channels_per_group,
+           X_rand_scale=st.floats(0.01, 1e3),
+           Y_scale=st.floats(0.2, 2.6),
+           Y_zero_point=st.integers(0, 5))
+    def test_instance_norm(self, shapes, torch_type, X_rand_scale,
```
Since these tests are almost identical can we write a single test that loops over different norm ops?

vkuzo replied:
will do this refactor in a separate follow-up

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized instancenorm operator"


Summary:

Adds a quantized instancenorm operator, which can reuse most of
groupnorm's logic.

Benchmarking shows that the quantized version is about 10x faster than
floating point for equivalent input sizes
(https://gist.github.com/vkuzo/2f230e84d26f26cc6030afdbfbc8e7f0)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107925](https://our.internmc.facebook.com/intern/diff/D21107925)

[ghstack-poisoned]
